blocks/library/cover-image/index.js

```diff
@@ -161,14 +169,18 @@ registerBlockType( 'core/cover-image', {
 	},
 
 	save( { attributes, className } ) {
-		const { url, title, hasParallax, hasBackgroundDim } = attributes;
+		const { url, title, hasParallax, dimRatio } = attributes;
 		const style = url
 			? { backgroundImage: `url(${ url })` }
 			: undefined;
-		const classes = classnames( className, {
-			'has-parallax': hasParallax,
-			'has-background-dim': hasBackgroundDim,
-		} );
+		const classes = classnames(
+			className,
+			dimRatioToClass( dimRatio ),
+			{
+				'has-background-dim': dimRatio !== 0,
```

Comment:
Was this added to try to preserve backwards compatibility with existing content? I'd have hoped it would be the case, but unfortunately existing cover image blocks appear to be flagged as invalid. The issue appears to be that the old content will have the `has-background-dim` class, but the `dimRatio` will not be found and therefore default to `0`, thus causing a subsequent save to save without `has-background-dim` and cause content loss (the validator in action!). Maybe we should assign the default for `dimRatio` to mimic what had previously existed as the default for `hasBackgroundDim` effect?

(For debugging validation, see also new #2837)

Reply:
> Was this added to try to preserve backwards compatibility with existing content?

Not really. For the issue of compatibility, I stopped to consider what we could do at the framework level. Regardless, something as small as this shouldn't be blocked by that, so I pushed 2884db5a621c832dc835cbd8f91bce0a6fd61c7c. Note, though, the weirdness of

```diff
 function dimRatioToClass( ratio ) {
-       return ratio === 0
+       return ( ratio === 0 || ratio === 50 )
                ? null
                : 'has-background-dim-' + Math.round( ratio / 10 );
 }
```

Since 50% dimming corresponds to the previous default value of `hasBackgroundDim`, we need to treat that specifically to avoid altering our block's class list.

Reply:
>Not really.

If not, why do we need `has-background-dim` at all?

Reply:
>Note, though, the weirdness of

I think this is fine, expected even. Similar to how we omit default values from serialized comments, we can omit the default value (50) from the class set.

Reply:
So... my answer was for the past only. :)

> I say this, though it may allow us to keep compatibility

Exactly. At first, it lingered while I pondered on the general compat stuff, but with that eab53ef it makes sense to keep it.